keps/2937-resource-transformer/README.md

```diff
@@ -0,0 +1,403 @@
+```yaml
+  kueue.x-k8s.io/accelerator-memory: 20G
+  nvidia.com/mig-1g.5gb: 2
+  nvidia.com/mig-2g.10gb: 1
+  memory: 100M
+  cpu: 1
+```
+
+Below is an example resource group that the cluster admin could define that
+prevents the usage of `7g.40gb` paritions and puts a tighter cap on `4g.20gb`
+partitions than would be implied by the `accelerator-memory` quota:
+```yaml
+...
+  - coveredResources: ["cpu", "memory", "kueue.x-k8s.io/accelerator-memory"]
+    flavors:
+    - name: default-flavor
+      resources:
+      - name: "cpu"
+        nominalQuota: ...
+      - name: "memory"
+        nominalQuota: ...
+      - name: kueue.x-k8s.io/accelerator-memory
+        nominalQuota: 200G
+      - name: nvidia.com/mig-1g.5gb
+        nominalQuota: 40  # 200G/5g
+      - name: nvidia.com/mig-2g.10gb
+        nominalQuota: 20  # 200G/10g
+      - name: nvidia.com/mig-4g.20gb
+        nominalQuota: 4  # less than the 10 that would be allowed by the acelerator-memory quota
+      - name: nvidia.com/mig-7g.40gb
+        nominalQuota: 0  # not allowed -- inefficient use of GPU resources
+```
+
+[mig]: https://docs.nvidia.com/datacenter/cloud-native/kubernetes/latest/index.html#the-mixed-strategy
+[instaslice]: https://github.com/openshift/instaslice-operator
+
+#### Story 2
+
+This KEP would allow Kueue to augment a Job's resource requests
+for quota purposes with additional resources that are not actually 
+available in the Nodes of the cluster. For example, it could be used to 
+define quotas in terms of a fictional currency `kueue.x-k8s.io/credits`
+to give "costs" to various resources. This could be used to approximate
+the actual monetary cost of a cloud resource (for example a larger or more powerful GPU
+could cost twice as many `credits` as a smaller or previous generation GPU).
+We'd expect in this usage pattern that `Replace` would be used so that
+resource list is just augmented with the computed `credits` resource.  We'd also
+expect that a ClusterQueue's `borrowingLimit` for `credits` would be set to `0`
+to enforce the intended fiscal constraint.
+
+## Design Details
+
+The only non-trivial design question is how best to specify the configuration.
+
+### Option 1: Singleton Matrix Input Focused
+The entire transformation matrix is defined in one object.  Most likely
+this object would be embedded inside Kueue's `Configuration` API, but
+it would also be possible to make it a separate top-level config API or 
+use a config map containing a single yaml field that would be deserialized
+into the type defined below.
+
+```go
+type ResourceTransformationStrategy string
+const Retain ResourceTransformationStrategy = "Retain"
+const Replace ResourceTransformationStrategy = "Replace"
+
+type ResourceTransformation struct {
+  // key is the output resource name and value is the quantity of the output resource 
+  // that corresponds to 1 unit of the input resource
+  Functions map[corev1.ResourceName ]resource.Quantity `json:"functions"`
```

Comment:
I find the `Functions` name confusing here. These are specifying the value of a single unit, which is then multiplied by the value of the input resource. Maybe it could be cleaner if called `UnitValues`?

IIUC the name `Functions` would be appropriate if we want to support some sort of placeholder functions, such as `5G * $count`, but this is out-of-scope IIUC.

Reply:
Agree `Functions` is fairly generic.  `UnitValues` could work.  `Conversions` is another possibility.

Reply:
I changed to Conversions for now.  Can change again if another term is preferred.

Reply:
`Conversions` sounds better than `Functions` for sure, and probably better than `UnitValues`. 

Still, the field doesn't specify the conversion formula, as this is hardcoded: `$baseValue * $count`. So, maybe another possibility: `conversionBaseValues`. Longer, but I think it's more descriptive. I'm open to other proposals.

Reply:
maybe `conversionFactors` Seems to fit how NIST [defines it](https://www.nist.gov/pml/owm/metric-si/unit-conversion#:~:text=A%20conversion%20factor%20is%20an,value%20expressed%20in%20different%20units.).

Reply:
IIUC it is more of an expression, say `1m=100cm`. It is close enough for me to be ok with it, but I'm slightly leaning towards `conversionBaseValues`.

Reply:
ok.  I'll go with `conversionBaseValues`.

Reply:
sgtm for the lack of a clearly better option for now